src/libraries/System.Data.Odbc/src/System/Data/Odbc/OdbcHandle.cs

```diff
@@ -201,8 +201,9 @@ internal ODBC32.SQLRETURN GetDiagnosticField(out string sqlState)
 
         internal ODBC32.SQLRETURN GetDiagnosticRecord(short record, out string sqlState, StringBuilder messageBuilder, out int nativeError, out short cchActual)
         {
-            // ODBC (MSDN) documents it expects a buffer large enough to hold 4(+L'\0') unicode characters
-            char[] buffer = new char[5];
+            // SQLGetDiagRecW expects a buffer large enough to hold a five-character state code plus a null-terminator
+            // See https://docs.microsoft.com/sql/odbc/reference/syntax/sqlgetdiagrec-function
+            char[] buffer = new char[6];
```

Comment:
How did we find this?  Did a test fail?

Reply:
Tests failed in https://github.com/dotnet/runtime/pull/69837.

Just went through how it was working before - we were ending up with an empty string as the state, since the parameter is not marked to be marshalled out, so it looks like the change to pin so that we get the data out exposed it.